File: tunasushi/src/main/java/com/tunasushi/tuna/TArc.java
package com.tunasushi.tuna;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;

/**
 * @author Tunasashimi
 * @date 2020-02-12 19:58
 * @Copyright 2020 TunaSashimi. All rights reserved.
 * @Description
 */
public class TArc extends TView {
    private Paint[] mPaints;
    private Paint mFramePaint;
    private boolean[] mUseCenters;
    private RectF[] mOvals;
    private RectF mBigOval;
    private float mStart;
    private float mSweep;
    private int mBigIndex;

    private static final float SWEEP_INC = 2;
    private static final float START_INC = 15;

    public TArc(Context context) {
        this(context, null);
    }

    public TArc(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TArc(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        tag = TArc.class.getSimpleName();

        mPaints = new Paint[4];
        mUseCenters = new boolean[4];
        mOvals = new RectF[4];

        mPaints[0] = new Paint();
        mPaints[0].setAntiAlias(true);
        mPaints[0].setStyle(Paint.Style.FILL);
        mPaints[0].setColor(0x88FF0000);
        mUseCenters[0] = false;

        mPaints[1] = new Paint(mPaints[0]);
        mPaints[1].setColor(0x8800FF00);
        mUseCenters[1] = true;

        mPaints[2] = new Paint(mPaints[0]);
        mPaints[2].setStyle(Paint.Style.STROKE);
        mPaints[2].setStrokeWidth(4);
        mPaints[2].setColor(0x880000FF);
        mUseCenters[2] = false;

        mPaints[3] = new Paint(mPaints[2]);
        mPaints[3].setColor(0x88888888);
        mUseCenters[3] = true;

        mBigOval = new RectF(40, 10, 280, 250);

        mOvals[0] = new RectF(10, 270, 70, 330);
        mOvals[1] = new RectF(90, 270, 150, 330);
        mOvals[2] = new RectF(170, 270, 230, 330);
        mOvals[3] = new RectF(250, 270, 310, 330);

        mFramePaint = new Paint();
        mFramePaint.setAntiAlias(true);
        mFramePaint.setStyle(Paint.Style.STROKE);
        mFramePaint.setStrokeWidth(0);
    }

    private void drawArcs(Canvas canvas, RectF oval, boolean useCenter, Paint paint) {
        canvas.drawRect(oval, mFramePaint);
        canvas.drawArc(oval, mStart, mSweep, useCenter, paint);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.drawColor(Color.WHITE);

        drawArcs(canvas, mBigOval, mUseCenters[mBigIndex], mPaints[mBigIndex]);

        for (int i = 0; i < 4; i++) {
            drawArcs(canvas, mOvals[i], mUseCenters[i], mPaints[i]);
        }

        mSweep += SWEEP_INC;
        if (mSweep > 360) {
            mSweep -= 360;
            mStart += START_INC;
            if (mStart >= 360) {
                mStart -= 360;
            }
            mBigIndex = (mBigIndex + 1) % mOvals.length;
        }
        invalidate();
    }
}

File: tunasushi/src/main/java/com/tunasushi/view/TArc.kt
package com.tunasushi.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import kotlin.jvm.JvmOverloads
import com.tunasushi.view.TView
import android.graphics.RectF
import android.util.AttributeSet
import com.tunasushi.view.TArc

/**
 * @author TunaSashimi
 * @date 2020-02-12 19:58
 * @Copyright 2020 TunaSashimi. All rights reserved.
 * @Description
 */
class TArc @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : TView(context, attrs, defStyleAttr) {
    private val mPaints: Array<Paint?>
    private val mFramePaint: Paint
    private val mUseCenters: BooleanArray
    private val mOvals: Array<RectF?>
    private val mBigOval: RectF
    private var mStart = 0f
    private var mSweep = 0f
    private var mBigIndex = 0
    private fun drawArcs(canvas: Canvas, oval: RectF?, useCenter: Boolean, paint: Paint?) {
        canvas.drawRect(oval!!, mFramePaint)
        canvas.drawArc(oval, mStart, mSweep, useCenter, paint!!)
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawColor(Color.WHITE)
        drawArcs(canvas, mBigOval, mUseCenters[mBigIndex], mPaints[mBigIndex])
        for (i in 0..3) {
            drawArcs(canvas, mOvals[i], mUseCenters[i], mPaints[i])
        }
        mSweep += SWEEP_INC
        if (mSweep > 360) {
            mSweep -= 360f
            mStart += START_INC
            if (mStart >= 360) {
                mStart -= 360f
            }
            mBigIndex = (mBigIndex + 1) % mOvals.size
        }
        invalidate()
    }

    companion object {
        private const val SWEEP_INC = 2f
        private const val START_INC = 15f
    }

    init {
        tag = TArc::class.java.simpleName
        mPaints = arrayOfNulls(4)
        mUseCenters = BooleanArray(4)
        mOvals = arrayOfNulls(4)
        mPaints[0] = Paint()
        mPaints[0]!!.isAntiAlias = true
        mPaints[0]!!.style = Paint.Style.FILL
        mPaints[0]!!.color = -0x77010000
        mUseCenters[0] = false
        mPaints[1] = Paint(mPaints[0])
        mPaints[1]!!.color = -0x77ff0100
        mUseCenters[1] = true
        mPaints[2] = Paint(mPaints[0])
        mPaints[2]!!.style = Paint.Style.STROKE
        mPaints[2]!!.strokeWidth = 4f
        mPaints[2]!!.color = -0x77ffff01
        mUseCenters[2] = false
        mPaints[3] = Paint(mPaints[2])
        mPaints[3]!!.color = -0x77777778
        mUseCenters[3] = true
        mBigOval = RectF(40f, 10f, 280f, 250f)
        mOvals[0] = RectF(10f, 270f, 70f, 330f)
        mOvals[1] = RectF(90f, 270f, 150f, 330f)
        mOvals[2] = RectF(170f, 270f, 230f, 330f)
        mOvals[3] = RectF(250f, 270f, 310f, 330f)
        mFramePaint = Paint()
        mFramePaint.isAntiAlias = true
        mFramePaint.style = Paint.Style.STROKE
        mFramePaint.strokeWidth = 0f
    }
}